Fix/dialog menu

zhss1983 opened the pull request:
Отрефакторил меню.
1. Теперь меню запускается по команде стразу после того как пользователь авторизуется.
2. Сделал заготовки в виде комментариев под остальные команды.
3. Сделал заготовки под функции меню.
4. Исправил ошибки в некоторых функциях-ответов на кнопки меню.
5. Исправил опечатку: ConreateAPIService  -> ConcreateAPIService

zhss1983 committed:
rebase

zhss1983 committed:
menu inserted in start_conversation, dialogs refactored.

zhss1983 committed:
Refactored start menu.

zhss1983 committed:
Refactored start menu.

zhss1983 committed:
Merge branch 'develop' into fix/dialog_menu

zhss1983 committed:
Update start_conversation.py

PEP8

zhss1983 commented:
После слияния с текущими изменениями перестала работать.
File "C:\Dev\ask_nenaprasno_bot\src\start_conversation.py", line 87, in is_expert_callback
context.user_data["user_id_in_trello"] = user_data.user_id_in_trello
AttributeError: 'UserData' object has no attribute 'user_id_in_trello'

AntonZelinsky commented on src/start_conversation.py:
```diff
@@ -137,42 +141,99 @@ async def menu(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
+# zayavki_handler = CommandHandler("zayavki", zayavki)
+# pravila_command_handler = CommandHandler("pravila", pravila)
+# help_unauthorized_command_handler = CommandHandler("help", help_unauthorized)
+# help_authorized_command_handler = CommandHandler("help", help_authorized)
```
![image](https://user-images.githubusercontent.com/1431279/176668401-ca7ee177-f7c0-4dbb-98d8-c130f1158a86.png)
Если весь этот код не реализуешь, то и не нужно оставлять закоментированные заглушки. Потом только вопросы будут вызывать. Причём и среди них видны проблемы.

zhss1983 replied:
Я предположил что это будут задачи которые будут решаться в будущем. Предположил что при таком варианте тем кто будет их выполнять было бы легче внести свой код. Раз это не так, я его удалю.

Альтернатива - подтвердите что я выполняю часть или всё из указанных задач по вызову помощи или запуску иных команд.

zhss1983 replied:
Отвечу по порядку:
Ты что-то про ConcreateAPIService, тут этого не вижу. - Я сделал комментарий в момент когда залил ветку. Затем посмотрел нет ли конфликтов, оказалось что есть и автоматически ветки не сольются. В процессе этого ConcreateAPIService превратилось в просто APIService.

Так сейчас работает код или нет? - В результате перехода к APIService фрагменты кода отвечающие за заглушки были удалены. Код запускается, но для авторизации необходим ответ от сервера. Он отсутствует. В результате не происходит авторизация, код падает. Переход к активации меню не происходит. То есть большая часть того функционала что я исправлял не работает.

Третья проблема с которой я столкнулся только что. 
Код не прошёл проверку PEP8:
./src/run_webhook_api.py:17:1: E303 too many blank lines (3)
./src/run_webhook_api.py:32:1: E303 too many blank lines (3)
./src/core/config.py:1:1: F401 'os' imported but unused
Проблема в том что в моём коде нет этих фрагментов. Нет ни лишних строк, ни импорта модуля OS. Они находятся в основной ветке develop и изменения каким-то образом попали именно туда. В результате не проходит проверка.

veselij replied:
что то у тебя не то, посмотри код в develop, там в core/config нет import os.
заглушка тоже есть.

zhss1983 replied:
Посмотри [PR](https://github.com/Studio-Yandex-Practicum/ask_nenaprasno_bot/pull/43) - там как раз параллельно эти моменты исправили. Только вот сомневаюсь что стоило сливать.

zhss1983 replied:
Проблемы совместимости исправил, PEP8 проходит, но пришлось добавить локально исключение W0703. Я не могу спрогнозировать какие именно исключения будут прилетать в декораторе логера.
Код запускается в полном объёме.

AntonZelinsky reviewed (approved):
Ты что-то про ConcreateAPIService, тут этого не вижу.

Так сейчас работает код или нет?

zhss1983 committed:
Added empty answer on push menu buttons.

zhss1983 committed:
Update run_webhook_api.py

PEP 8
Fix
./src/run_webhook_api.py:17:1: E303 too many blank lines (3)
./src/run_webhook_api.py:33:1: E303 too many blank lines (3)

zhss1983 committed:
Update run_webhook_api.py

PEP 8
Fix
./src/run_webhook_api.py:17:1: E303 too many blank lines (3)
./src/run_webhook_api.py:33:1: E303 too many blank lines (3)

zhss1983 committed:
Update run_webhook_api.py

PEP 8
Fix
./src/run_webhook_api.py:17:1: E303 too many blank lines (3)
./src/run_webhook_api.py:33:1: E303 too many blank lines (3)

zhss1983 committed:
Try resolve PEP8 problems.

zhss1983 committed:
Fix empty text problem.

zhss1983 committed:
Fix empty text problem.

zhss1983 committed:
Rebased with develop branch.

zhss1983 committed:
Fix empty text problem.

zhss1983 committed:
Fix empty text problem.

zhss1983 committed:
Merge branch 'develop' into fix/dialog_menu

zhss1983 committed:
Fixed dialog menu.

veselij commented on src/start_conversation.py:
```diff
@@ -81,10 +81,9 @@ async def is_expert_callback(update: Update, context: CallbackContext):
     user_data = await api_service.authenticate_user(telegram_id=telegram_id)
     if user_data is None:
-        await update.callback_query.edit_message_text(text="Ошибка авторизации")
+        await update.callback_query.callback_query.edit_message_text(text="Ошибка авторизации")
```
callback_data не хватате, бот падает

veselij commented on src/start_conversation.py:
```diff
@@ -137,42 +140,91 @@ async def menu(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
 menu_command_handler = CommandHandler("menu", menu)
 
-callback_menu_handler = CallbackQueryHandler(handling_menu_button_click_callback)
+authorized_user_command_handlers = (menu_command_handler,)
+
+unauthorized_user_command_handlers = []
```
а зачем нам лист (к тому же пустой) и тапл, подставить просто menu_command_handler

zhss1983 replied:
Да. Я сейчас чай пил и тоже подумал - зачем список распаковывать если там в будущем один единственный хэндлер предпологается. Затем к компу и вернулся. А тут...

veselij commented on src/start_conversation.py:
```diff
@@ -137,42 +140,91 @@ async def menu(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
                 support_or_consult_callback, pattern=callback_data.CALLBACK_SUPPORT_OR_CONSULT_COMMAND
             ),
         ],
         states.NEW_EXPERT_STATE: [CallbackQueryHandler(after_registr_message_callback)],
```
в данном стейте надо отправить пользователю через callback сообщение со ссылкой на анкету для того чтобы стать экспертом, а не приверженное сообщение, которое надо отправить после авторизации и перед установкой тайм зоны

veselij commented on src/start_conversation.py:
```diff
@@ -137,42 +140,91 @@ async def menu(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
+            *unauthorized_user_command_handlers,
             CallbackQueryHandler(registr_as_expert_callback, pattern=callback_data.CALLBACK_REGISTR_AS_EXPERT_COMMAND),
             CallbackQueryHandler(
                 support_or_consult_callback, pattern=callback_data.CALLBACK_SUPPORT_OR_CONSULT_COMMAND
```
в support_or_consult_callback надо отправит сообщение со ссылкой на то как помочь организации

veselij commented on src/start_conversation.py:
```diff
@@ -137,42 +140,91 @@ async def menu(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
         states.NEW_EXPERT_STATE: [CallbackQueryHandler(after_registr_message_callback)],
         states.TIMEZONE_STATE: [
+            *authorized_user_command_handlers,
             CallbackQueryHandler(timezone_callback, pattern=callback_data.CALLBACK_TIMEZONE_COMMAND),
             CallbackQueryHandler(skip_timezone_callback, pattern=callback_data.CALLBACK_SKIP_TIMEZONE_COMMAND),
             CallbackQueryHandler(timezone_message_callback),
```
сюда надо встроить выбор таймзоны. PR с таймзонами смерджили, посмотри пожалуйста

zhss1983 replied:
Я соединил get_timezone и start_conversation. Для нормальной работы в структуре меню пришлось частично переделать код (добавил сообщений и стейтов). Убрал не используемые стейты из констант и из start_conversation, убрал с ними и callback функции.

zhss1983 committed:
Fixed dialog menu.

zhss1983 committed:
Added docstrings.

zhss1983 committed:
Deleted unused constants.

zhss1983 committed:
Deleted unused constants.

kapkaevandrey reviewed (approved):
Некоторые стилистические темы мне не нравятся, но и code style устанавливаю не я )))) так что всё норм.

veselij reviewed (changes requested):
тут из-за смешивания изменений стилей и форматирования очень сложно понять, что поменялось. 
давай разобьем на 2 PR. в этом пофиксим только стили. а второрым будем суть править. на счет сути, точно не надо делать списки хендлеров и распаковывать их в меню - так как постоянно скролить смотреть что-это. и второе, ты все из таймзоны перенес сюда, причем оставив какой-то непонятный фрагмент в файле get_timezone, почему не получилось все в get_timezone оставить?

zhss1983 commented:
Разделил меню и диалоги на блоки. Вход в блок авторизации и выбора timezone, затем переход на блок меню и команд. Реакция на команды и меню разделена тематически на отдельные файлы. Реализованы шаблоны для всех команд. Из блока выбора timezone вынес бизнес логику. Изменил константы таким образом чтобы при их печати было ясно что это константы. Добавил меню бота, дописал туда недостающие команды. Меню имеет два уровня, для авторизованных  и для неавторизованных пользователей.

MatMerd commented on src/start_conversation.py:
```diff
@@ -1,13 +1,85 @@
+    return states.MENU_STATE
+
+
+async def get_timezone_from_text_message_callback(update: Update, context: CallbackContext):
```
Я думал функции, связанные с таймзонами, ты вчера хотел перенести в отдельный файл.

MatMerd commented:
К самим conversation вопросов нет. Но я думал при рефакторе будут разделены клавиатура и функционал для таймзон

zhss1983 committed:
Rebased with develop branch.

zhss1983 committed:
Rebase continue

zhss1983 committed:
Rebase continue

zhss1983 committed:
Rebase continue

zhss1983 commented:
Прошу прощения - придётся подождать. Всё было. Умудрился снести файлы, да так что восстановить удалось всего 2 из примерно 8. Всё было качественно разнесено, понятно. Сейчас спешно восстанавливаю. Думаю ночью закончу.

zhss1983 committed:
Промежуточная точка

zhss1983 committed:
Промежуточная точка

zhss1983 committed:
Divided the menus and dialogs into blocks. Log in to the authorization and timezone selection block, then go to the menu and command block. The reaction to commands and menus is divided thematically into separate files. Templates for all commands are implemented. Timezone took out the business logic from the selection block. I changed the constants so that when they were printed it was clear that they were constants. Added the bot menu, added the missing commands there. The menu has two levels, for authorized and for unauthorized users.

zhss1983 committed:
Divided the menus and dialogs into blocks. Log in to the authorization and timezone selection block, then go to the menu and command block. The reaction to commands and menus is divided thematically into separate files. Templates for all commands are implemented. Timezone took out the business logic from the selection block. I changed the constants so that when they were printed it was clear that they were constants. Added the bot menu, added the missing commands there. The menu has two levels, for authorized and for unauthorized users.

zhss1983 committed:
Divided the menus and dialogs into blocks. Log in to the authorization and timezone selection block, then go to the menu and command block. The reaction to commands and menus is divided thematically into separate files. Templates for all commands are implemented. Timezone took out the business logic from the selection block. I changed the constants so that when they were printed it was clear that they were constants. Added the bot menu, added the missing commands there. The menu has two levels, for authorized and for unauthorized users.

zhss1983 committed:
Divided the menus and dialogs into blocks. Log in to the authorization and timezone selection block, then go to the menu and command block. The reaction to commands and menus is divided thematically into separate files. Templates for all commands are implemented. Timezone took out the business logic from the selection block. I changed the constants so that when they were printed it was clear that they were constants. Added the bot menu, added the missing commands there. The menu has two levels, for authorized and for unauthorized users.

zhss1983 committed:
Divided the menus and dialogs into blocks. Log in to the authorization and timezone selection block, then go to the menu and command block. The reaction to commands and menus is divided thematically into separate files. Templates for all commands are implemented. Timezone took out the business logic from the selection block. I changed the constants so that when they were printed it was clear that they were constants. Added the bot menu, added the missing commands there. The menu has two levels, for authorized and for unauthorized users.

zhss1983 committed:
Divided the menus and dialogs into blocks. Log in to the authorization and timezone selection block, then go to the menu and command block. The reaction to commands and menus is divided thematically into separate files. Templates for all commands are implemented. Timezone took out the business logic from the selection block. I changed the constants so that when they were printed it was clear that they were constants. Added the bot menu, added the missing commands there. The menu has two levels, for authorized and for unauthorized users.

zhss1983 committed:
Divided the menus and dialogs into blocks. Log in to the authorization and timezone selection block, then go to the menu and command block. The reaction to commands and menus is divided thematically into separate files. Templates for all commands are implemented. Timezone took out the business logic from the selection block. I changed the constants so that when they were printed it was clear that they were constants. Added the bot menu, added the missing commands there. The menu has two levels, for authorized and for unauthorized users.

zhss1983 commented:
@veselij Я закончил восстановление.
Разделил меню и диалоги на блоки. Вход в блок авторизации и выбора timezone, затем переход на блок меню и команд. Реакция на команды и меню разделена тематически на отдельные файлы. Реализованы шаблоны для всех команд. Из блока выбора timezone вынес бизнес логику. Изменил константы таким образом чтобы при их печати было ясно что это константы. Добавил меню бота, дописал туда недостающие команды. Меню имеет два уровня, для авторизованных и для неавторизованных пользователей.

zhss1983 committed:
Divided the menus and dialogs into blocks. Log in to the authorization and timezone selection block, then go to the menu and command block. The reaction to commands and menus is divided thematically into separate files. Templates for all commands are implemented. Timezone took out the business logic from the selection block. I changed the constants so that when they were printed it was clear that they were constants. Added the bot menu, added the missing commands there. The menu has two levels, for authorized and for unauthorized users.

zhss1983 committed:
Divided the menus and dialogs into blocks. Log in to the authorization and timezone selection block, then go to the menu and command block. The reaction to commands and menus is divided thematically into separate files. Templates for all commands are implemented. Timezone took out the business logic from the selection block. I changed the constants so that when they were printed it was clear that they were constants. Added the bot menu, added the missing commands there. The menu has two levels, for authorized and for unauthorized users.

zhss1983 commented:
Исправил маленький баг.
@zhss_test_bot

zhss1983 committed:
Divided the menus and dialogs into blocks. Log in to the authorization and timezone selection block, then go to the menu and command block. The reaction to commands and menus is divided thematically into separate files. Templates for all commands are implemented. Timezone took out the business logic from the selection block. I changed the constants so that when they were printed it was clear that they were constants. Added the bot menu, added the missing commands there. The menu has two levels, for authorized and for unauthorized users.

zhss1983 committed:
Divided the menus and dialogs into blocks. Log in to the authorization and timezone selection block, then go to the menu and command block. The reaction to commands and menus is divided thematically into separate files. Templates for all commands are implemented. Timezone took out the business logic from the selection block. I changed the constants so that when they were printed it was clear that they were constants. Added the bot menu, added the missing commands there. The menu has two levels, for authorized and for unauthorized users.

zhss1983 commented:
Хотел бы обратить внимание что при работе через telegram-desktop попытка отправить геоданные вызывает ошибку в логике работы программы и незначительно меняет выводимый текст. Не критично, но я не смог разобраться что с этим делать и как пофиксить.

zhss1983 committed:
Divided the menus and dialogs into blocks. Log in to the authorization and timezone selection block, then go to the menu and command block. The reaction to commands and menus is divided thematically into separate files. Templates for all commands are implemented. Timezone took out the business logic from the selection block. I changed the constants so that when they were printed it was clear that they were constants. Added the bot menu, added the missing commands there. The menu has two levels, for authorized and for unauthorized users.

zhss1983 committed:
Added loging to all methods which work with telegram.

veselij commented on src/constants/commands.py:
```diff
@@ -0,0 +1,8 @@
+GET_TIMEZONE = "get_timezone"
+DEADLINES = "srok"
+STATEMENTS = "zayavki"
+SERVICE_POLICE = "pravila"
```
POLICY

zhss1983 replied:
Политика сервиса, правила сервиса - как-то так. Значение полиция в данном случае не актуально.

zhss1983 commented:
Я не понимаю откуда взялась эта старая версия. Этого нет в ветке.
![image](https://user-images.githubusercontent.com/83511510/177056720-9ea82a70-fc57-4439-b70b-9cb1a90e6f58.png)

zhss1983 committed:
Rebase continue

zhss1983 committed:
Deleted unused methods (get_timezone.get_timezone)

zhss1983 committed:
replace states.END_STATE on ConversationHandler.END

zhss1983 committed:
Refactoring

zhss1983 commented:
> тут из-за смешивания изменений стилей и форматирования очень сложно понять, что поменялось.
> давай разобьем на 2 PR. в этом пофиксим только стили. а второрым будем суть править. на счет сути, точно не надо делать списки хендлеров и распаковывать их в меню - так как постоянно скролить смотреть что-это. и второе, ты все из таймзоны перенес сюда, причем оставив какой-то непонятный фрагмент в файле get_timezone, почему не получилось все в get_timezone оставить?

Это ещё актуально? Я только что заметил этот комент. До этого умудрился проморгать.

zhss1983 commented:
Код из файла timezone переиспользуется в двух ConversationHandler. Его не получится распаковать в отдельный. Остальные строго по темам разбиты.

zhss1983 committed:
Merge branch 'develop' into fix/dialog_menu

zhss1983 committed:
fix name menu_comands -> menu_commands

zhss1983 commented:
Я предполагаю что в комментарии выше речь шла о старой версии когда у меня всё было свалено в файл start_conversation. Сейчас это всё вынесено для более лёгкого восприятия. Однако поясню для понимания:

>ты все из таймзоны перенес сюда, причем оставив какой-то непонятный фрагмент в файле get_timezone, почему не получилось все в get_timezone оставить?

В src.get_timezone остались функции отвечающие за непосредственное определение таймзоны, а всё что касается диалога с пользователем я перенёс к себе в src.conversation.timezone

> не надо делать списки хендлеров и распаковывать их в меню

Могу не делать, но придётся нарушать DRY, считаете это приемлемым? У меня в menu_commands распаковывается только словарь states_timezone_conversation_dict (это весь функционал timezone.py - полное переиспользование, только без переноса всего кода из timezone.py) и список команд authorized_user_command_handlers (это опять же все команды которые применяются в 2 местах).

Применяемые имена соответствуют соглашениям. Где возможно - код переиспользован по максимуму. Логические блоки разбиты по файлам, команды каждого типа в своём файле и для работы с ними необходимо всего в 2 местах вписать свой код - реакцию на вызов команды текстом через / и раекцию на нажатие кнопки меню. Они отличаются прилетающими update, по этой причине я не стал их объединять (в противном случае возможно долго ловить ошибки в зависимости от того откуда будет следовать вызов).

Для команд статистики это src.conversation.statistic
Для запросов (просроченных и актуальных) src.conversation.requests
Для правил сервиса src.conversation.service_police, там минимальный функционал уже прописан.
Для работы с таймзоной - src.conversation.timezone - там код так же прописан, но "бизнес логика" вынесена, осталась в src.get_timezone

При формировании структуры я исходил из удобства восприятия и концепций "чистого кода" в том объёме в котором я успел их воспринять.
Мне крайне не хочется снова сводить всё в один файл так как скроллинг по нему занимает куда больше времени чем переключение между вкладками.

Если я не прав, просьба не молчать в ожидание когда я допру, а объяснить хотя бы куда смотреть и/или что не нравится.

zhss1983 committed:
Настроил poetry, но в файлы зависимостей не была вписана одна библиотека. Я её дописал. Без неё код не стартует. Ещё дописал flake8 с флагом --dev для проверки на соответствие PEP8

MatMerd commented:
Сергей, я писал тебе в личку, напишу тут. Мы не можем смерджить такой большой коммит. Если можешь, можно созвониться и мы поговорим об этом

zhss1983 committed:
Исправил замечания.
